src/common/framework/test_group.ts

```diff
@@ -50,33 +50,43 @@ export class TestGroup<F extends Fixture> implements RunCaseIterable {
   }
 
   // TODO: This could take a fixture, too, to override the one for the group.
-  test(name: string, fn: TestFn<F>): Test<F> {
+  test(name: string): TestBuilderWithName<F, never> {
     // Replace spaces with underscores for readability.
     assert(name.indexOf('_') === -1, 'Invalid test name ${name}: contains underscore (use space)');
     name = name.replace(/ /g, '_');
 
     this.checkName(name);
 
-    const test = new Test<F>(name, this.fixture, fn);
+    const test = new TestBuilder<F, never>(name, this.fixture);
     this.tests.push(test);
     return test;
   }
 }
 
-// This test is created when it's inserted, but may be parameterized afterward (.params()).
-class Test<F extends Fixture> {
-  readonly name: string;
-  readonly fixture: FixtureClass<F>;
-  readonly fn: TestFn<F>;
+interface TestBuilderWithName<F extends Fixture, P extends {}> extends TestBuilderWithParams<F, P> {
+  params<NewP extends {}>(specs: Iterable<NewP>): TestBuilderWithParams<F, NewP>;
+}
+
+interface TestBuilderWithParams<F extends Fixture, P extends {}> {
+  fn(fn: TestFn<F, P>): void;
+}
+
+class TestBuilder<F extends Fixture, P extends {}> {
```

Comment:
Could probably work, but I wanted to make the changes be as purely at the type level as possible (not actually changing the behavior). I'll consider how hard this would be

Reply:
Oh, this is actually not possible. We have to use the same object through the builder chain, because `g.test()` is what registers the test in the first place - we don't register it at the end. If we returned new objects during the chain, `g` wouldn't know about them.

Reply:
Ah, okay thanks for the clarification